src/PermissionProvider/ConfigPermissions.php

```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace Drupal\entity_sync\PermissionProvider;
+
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+
+/**
+ * Provides permissions for each entity_sync.sync operation.
+ */
+class ConfigPermissions {
+
+  use StringTranslationTrait;
+
+  /**
+   * Returns an array of permissions for entity_sync.sync configs.
+   *
+   * @return array
+   *   The entity_sync.sync permissions.
+   *   @see \Drupal\user\PermissionHandlerInterface::getPermissions()
+   */
+  public function getPermissions() {
+    $permissions = [];
+
+    $config_factory = \Drupal::configFactory();
+
+    // Loop through each entity_sync.sync configurations and create
+    // corresponding permissions.
+    foreach (
+      $config_factory->listAll('entity_sync.sync.') as $config_name
+    ) {
+      $config = $config_factory->get($config_name)->get();
+
+      // Do not proceed if operations are not set for config.
+      if (!is_array($config['operations'])) {
+        continue;
+      }
+
+      $bundle = $config['entity']['bundle'];
+      $type_id = $config['entity']['type_id'];
+      $type_id_label = ucfirst($type_id);
+
+      // Loop through each entity sync operation and create permission.
+      foreach ($config['operations'] as $operation) {
+        $operation_id = $operation['id'];
+
+        $permissions += [
+          "$type_id $bundle $operation_id" => [
+            'title' => $this->t(
+              '%type_id_label: @bundle @operation',
+              [
+                '%type_id_label' => $type_id_label,
+                '@bundle' => $bundle,
+                '@operation' => $operation['label'],
+              ]
+            ),
+          ],
+        ];
```

Comment:
**CHANGE REQUEST**

You need to properly load the labels; load the entity type using the entity type manager.

**CHANGE REQUEST**

The permissions should be:
* Entity type permissions: `entity_sync ${OPERATION_ID} ${ENTITY_TYPE_ID}`.
* Bundle permissions: `entity_sync ${OPERATION_ID} ${BUNDLE} ${ENTITY_TYPE_ID}`.

The title should be:
* Entity type permissions: `Run the @operation_label operation on @entity_type_plural_label`.
* Bundle permissions: `@bundle: Run the @operation_label operation on @entity_type_plural_label`.

Reply:
I did not clearly understood this. Could you please take a look into what I did. Thanks!